examples/manifold_gradients.dx

```diff
@@ -0,0 +1,342 @@
+  -- Must satisfy tangentExponentialMap p zero == p
+  -- (tangentExponentialMap p) will be linearized around zero, so it is
+  -- important that any operations used have well-defined derivatives for inputs
+  -- p, zero.
+  tangentExponentialMap: a -> b -> a
+  -- (tangentLogarithmMap p) is the pointwise inverse to
+  -- (tangentExponentialMap p).
+  -- Must satisfy tangentLogarithmMap p p == zero
+  -- (tangentLogarithmMap p) will be linearized around p, so it is important
+  -- that any operations used have well-defined derivatives for inputs p, p.
+  tangentLogarithmMap: a -> a -> b
+
+
+'### Differentiation functions
+
+'We can then define a version of the Dex `linearize` function which works on
+manifolds.  To see intuitively why this works, consider the following diagram.
+We would like to obtain a linearized version of the function "across the top" of
+the diagram, and we can form this by linearizing the function formed from the
+bottom three sides of the square.
+
+:html
+   "<img
+     src=\"https://user-images.githubusercontent.com/33986545/122416078-5a23ef00-cf80-11eb-83d5-d0d5986e8d20.png\"
+     width=\"100%\">"
+
+
+def manifoldLinearize [TangentSpace a a', TangentSpace b b', Add a']
+    (f: a -> b) (x: a) : (b & a' --o b') = 
+  y = f x
+  tangentSpaceMapping : a' -> b' =
+    \ v . tangentLogarithmMap y $ f $ tangentExponentialMap x v
+  rawLinearize = linearize tangentSpaceMapping zero
+  (y, snd rawLinearize)
+
+' A full proof that this is in fact a correct formulation is a bit more
+involved, so we leave it out here. The general idea is to use the differentiable
+curves formulation of the tangent space, and to treat the tangent space at a 
+point $x$ as both the tangent space, and as a local chart.
+
+' Before we can define `grad` for manifolds, we need to create the trivial
+instance of a TangentSpace for vector spaces. In this case, we view the tangent
+space at $x$ to be just a copy of the vector space, centred on $x$.
+
+instance [VSpace a] TangentSpace a a
+  tangentExponentialMap = \ p v . p + v
+  tangentLogarithmMap = \ p q . q - p
+
+' Now we can define a version of the Dex `grad` function which works on
+manifolds.
+
+def manifoldGrad [TangentSpace a a', VSpace a'] (f: a -> Float) (x: a) : a' = 
+  linearized: (Float & (a' --o Float)) = manifoldLinearize f x
+  transposeLinear (snd linearized) 1.0
+
+
+' ### Equivalence to standard differentiation
+As a sense check we make sure that these functions give the same output as the
+non-manifold versions on functions which are defined on $\mathbb{R}^n$.
+
+x = for i:(Fin 10). IToF (ordinal i) / 10.
+
+def myFunc (x : (n => Float)) : Float =
+  sum $ for i. if (mod (ordinal i) 2 == 0) then (exp x.i) else (sin x.i)
+
+original = linearize myFunc x
+-- Need to explicitly set up the types here, due to missing associated types.
+new : (Float & ((Fin 10 => Float) --o Float)) = manifoldLinearize myFunc x
+
+fst original
+fst new
+
+snd original x
+snd new x
```

Comment:
I think that you should be able to say something like `original ~~ new` (and see the output be `True`). In fact, if you say `linearize myFunc x ~~ manifoldLinearize myFunc x` that might give just enough information to type inference to avoid this extra annotation!

Reply:
This doesn't seem to work, I guess since `snd original` is a function and so can't be checked for (near) equality? I've added the tests separately on `fst` and `snd` but the type annotations still seem to be required.

I also found a small bug in the implementation of `HasAllClose` for array types in the prelude. I've added a fix here but let me know if you'd prefer it split out to a separate PR.